bugsnag-android-core/src/main/java/com/bugsnag/android/StorageModule.kt

```diff
@@ -2,56 +2,72 @@ package com.bugsnag.android
 
 import android.content.Context
 import com.bugsnag.android.internal.BackgroundTaskService
+import com.bugsnag.android.internal.BugsnagStoreMigrator.migrateLegacyFiles
 import com.bugsnag.android.internal.ImmutableConfig
 import com.bugsnag.android.internal.TaskType
-import com.bugsnag.android.internal.dag.DependencyModule
+import com.bugsnag.android.internal.dag.BackgroundDependencyModule
+import com.bugsnag.android.internal.dag.Provider
 
 /**
  * A dependency module which constructs the objects that store information to disk in Bugsnag.
  */
 internal class StorageModule(
     appContext: Context,
-    immutableConfig: ImmutableConfig,
+    private val immutableConfig: ImmutableConfig,
     bgTaskService: BackgroundTaskService
-) : DependencyModule() {
+) : BackgroundDependencyModule(bgTaskService, TaskType.IO) {
 
-    val sharedPrefMigrator = bgTaskService.submitTask(
-        TaskType.IO,
+    val bugsnagDir = provider {
+        migrateLegacyFiles(immutableConfig.persistenceDirectory)
+    }
+
+    val sharedPrefMigrator = provider {
         SharedPrefMigrator(appContext)
-    )
+    }
 
-    val deviceIdStore = bgTaskService.submitTask(
-        TaskType.IO,
+    val deviceIdStore = provider {
         DeviceIdStore(
             appContext,
             sharedPrefMigrator = sharedPrefMigrator,
             logger = immutableConfig.logger,
             config = immutableConfig
         )
-    )
+    }
 
-    val userStore = UserStore(
-        immutableConfig,
-        deviceIdStore,
-        sharedPrefMigrator = sharedPrefMigrator,
-        logger = immutableConfig.logger
-    )
+    val userStore = provider {
+        UserStore(
+            immutableConfig.persistUser,
+            bugsnagDir,
+            deviceIdStore.map { it.load() },
+            sharedPrefMigrator = sharedPrefMigrator,
+            logger = immutableConfig.logger
+        )
+    }
 
-    val lastRunInfoStore = LastRunInfoStore(immutableConfig)
+    val lastRunInfoStore = provider {
+        LastRunInfoStore(immutableConfig)
+    }
 
-    val sessionStore =
+    val sessionStore = provider {
         SessionStore(
-            immutableConfig,
+            bugsnagDir.get(),
+            immutableConfig.maxPersistedSessions,
+            immutableConfig.apiKey,
             immutableConfig.logger,
             null
         )
+    }
 
-    val lastRunInfo = lastRunInfo()
-
-    private fun lastRunInfo(): LastRunInfo? {
+    val lastRunInfo = lastRunInfoStore.map { lastRunInfoStore ->
         val info = lastRunInfoStore.load()
         val currentRunInfo = LastRunInfo(0, crashed = false, crashedDuringLaunch = false)
         lastRunInfoStore.persist(currentRunInfo)
-        return info
+        return@map info
+    }
+
+    fun loadUser(initialUser: User): Provider<UserState> = provider {
```

Comment:
This appears to be doubly-wrapped in providers - what's the reason for UserStore being handled differently?

Reply:
Discussed IRL - other fields in client will probably go this way in future changes.